athena-synapse/src/main/java/com/amazonaws/athena/connectors/synapse/SynapseMetadataHandler.java

```diff
@@ -295,36 +299,131 @@ private Schema getSchema(Connection jdbcConnection, TableName tableName, Schema
-        try (ResultSet resultSet = getColumns(jdbcConnection.getCatalog(), tableName, jdbcConnection.getMetaData());
-             Connection connection = getJdbcConnectionFactory().getConnection(getCredentialProvider());
+        SchemaBuilder schemaBuilder;
+        HashMap<String, List<String>> columnNameAndDataTypeMap = new HashMap<>();
+        Matcher m = Pattern.compile("([a-zA-Z]+)://([^;]+);(.*)").matcher(jdbcConnection.getMetaData().getURL());
+        String hostName = "";
+
+        if (m.find() && m.groupCount() == 3) {
+            hostName = m.group(2);
+        }
+
+        try (Connection connection = getJdbcConnectionFactory().getConnection(getCredentialProvider());
              PreparedStatement stmt = connection.prepareStatement(dataTypeQuery)) {
-            // fetch data types of columns and prepare map with column name and datatype.
+            // fetch data types of columns and prepare map with column name and datatype information.
             stmt.setString(1, tableName.getSchemaName() + "." + tableName.getTableName());
             try (ResultSet dataTypeResultSet = stmt.executeQuery()) {
                 while (dataTypeResultSet.next()) {
-                    dataType = dataTypeResultSet.getString("DATA_TYPE");
-                    columnName = dataTypeResultSet.getString("COLUMN_NAME");
-                    hashMap.put(columnName.trim(), dataType.trim());
+                    List<String> columnDetails = List.of(
+                            dataTypeResultSet.getString("DATA_TYPE").trim(),
+                            dataTypeResultSet.getString("PRECISION").trim(),
+                            dataTypeResultSet.getString("SCALE").trim());
+                    columnNameAndDataTypeMap.put(dataTypeResultSet.getString("COLUMN_NAME").trim(), columnDetails);
                 }
             }
+        }
+
+        if (StringUtils.isNotBlank(hostName) && hostName.contains("ondemand")) {
+            // getColumns() method from SQL Server driver is causing an exception in case of Azure Serverless environment.
+            // so doing explicit data type conversion
+            schemaBuilder = doDataTypeConversion(columnNameAndDataTypeMap, tableName.getSchemaName());
+        }
+        else {
+            schemaBuilder = doDataTypeConversionForNonCompatible(jdbcConnection, tableName, columnNameAndDataTypeMap);
+        }
+        // add partition columns
+        partitionSchema.getFields().forEach(schemaBuilder::addField);
+        return schemaBuilder.build();
+    }
+
+    private SchemaBuilder doDataTypeConversion(HashMap<String, List<String>> columnNameAndDataTypeMap, String schemaName)
+    {
+        String columnName;
+        List<String> dataTypeDetails;
+        String dataType;
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        // Initializing varchar type as default
+        ArrowType columnType = Types.MinorType.VARCHAR.getType();
```

Comment:
Additionally, this won't work as intended, because the default for the `n+1` iteration will be whatever the value was set to in the `nth` iteration.

Fixing the variable scope should also fix this problem.

Reply:
agreed.